### PR TITLE
Lease hash fix

### PR DESCRIPTION
**Pull request by fanminshi**
functional-tester: add retries to hash checking

allows hashes to converge through retrying.

### CODE REVIEW

File: tools/functional-tester/etcd-tester/lease_stresser.go [hunk 1]
```diff
@@ -211,13 +211,14 @@ func (ls *leaseStresser) createLeases() {
...
 			if err != nil {
-				plog.Errorf("lease creation error: (%v)", err)
+				plog.Debugf("lease creation error: (%v)", err)
 				return
 			}
 			plog.Debugf("lease %v created", leaseID)
 			// if attaching keys to the lease encountered an error, we don't add the lease to the aliveLeases map
 			// because invariant check on the lease will fail due to keys not found
 			if err := ls.attachKeysWithLease(leaseID); err != nil {
+				plog.Infof("unable to attach keys to lease %d error (%v)", leaseID, err)
```
Comment (xiang90) [inline tools/functional-tester/etcd-tester/lease_stresser.go]: is this expected or unexpected? will we issue this rpc during failure injecting period?

File: tools/functional-tester/etcd-tester/lease_stresser.go [hunk 2]
```diff
@@ -239,6 +240,7 @@ func (ls *leaseStresser) randomlyDropLeases() {
 			// if randomlyDropLease encountered an error such as context is cancelled, remove the lease from aliveLeases
 			// becasue we can't tell whether the lease is dropped or not.
 			if err != nil {
+				plog.Errorf("drop lease %v has failed error (%v)", leaseID, err)
```
Comment (xiang90) [inline tools/functional-tester/etcd-tester/lease_stresser.go]: is this expected or unexpected? will we issue this rpc during failure injecting period?

File: tools/functional-tester/etcd-tester/lease_stresser.go (hunk 1 above)
Comment (fanminshi) [inline tools/functional-tester/etcd-tester/lease_stresser.go]: the failure can happen during failure injecting period. most of failures are expected. I'll change it to debug.

File: tools/functional-tester/etcd-tester/checks.go [hunk 3]
```diff
@@ -22,6 +22,8 @@ import (
 	"golang.org/x/net/context"
 )
 
+const retries = 3
```
Comment (xiang90) [inline tools/functional-tester/etcd-tester/checks.go]: we need to check both rev and hash. we are able to do it anyway. then we can know that: OH... we thought rev is stablized but it is not... it caused hash to be unstable :(.

Comment (fanminshi) [inline tools/functional-tester/etcd-tester/checks.go]: I see. make sense.

**Comment by fanminshi** [comment]
All fixed. PTAL cc @xiang90 @gyuho

**Comment by gyuho** [comment]
lgtm. defer to @xiang90 

@fanminshi Did you try running this? Or let's deploy this on 5-node cluster after merge.

File: tools/functional-tester/etcd-tester/checks.go [hunk 4]
```diff
@@ -39,39 +41,42 @@ func newHashChecker(hrg hashAndRevGetter) Checker { return &hashChecker{hrg} }
...
 		if err != nil {
 			plog.Printf("#%d failed to get current revisions (%v)", i, err)
 			continue
 		}
-		if _, ok = getSameValue(revs); ok {
-			break
-		}
 
-		plog.Printf("#%d inconsistent current revisions %+v", i, revs)
+		_, sameRev = getSameValue(revs)
```
Comment (xiang90) [inline tools/functional-tester/etcd-tester/checks.go]: i think we also want to detect that they are stablized. which means there is no changes for the last N seconds.

Comment (fanminshi) [inline tools/functional-tester/etcd-tester/checks.go]: Added the stabilization logic.

**Comment by fanminshi** [comment]
All fixed. PTAL @xiang90 

@gyuho I run the code on my local machines and it runs. I think we should deploy it to 5 nodes cluster after it gets merged.

File: tools/functional-tester/etcd-tester/checks.go [hunk 5]
```diff
@@ -22,6 +22,11 @@ import (
 	"golang.org/x/net/context"
 )
 
+const (
+	retries                     = 3
+	sleepForStabliztionInSecond = 1 * time.Second
```
Comment (xiang90) [inline tools/functional-tester/etcd-tester/checks.go]: stabilizationPeriod? probably change to 3 seconds?

File: tools/functional-tester/etcd-tester/checks.go [hunk 6]
```diff
@@ -39,39 +44,81 @@ func newHashChecker(hrg hashAndRevGetter) Checker { return &hashChecker{hrg} }
...
-		ok     bool
-	)
-	// retry in case of transient failure
-	for i := 0; i < 3; i++ {
-		revs, hashes, err = hc.hrg.getRevisionHash()
+func (hc *hashChecker) checkRevAndHashes() (err error) {
+	// retries in case of transient failure or etcd nodes have not stablized yet.
+	for i := 0; i < retries; i++ {
+		revsStable, err := hc.areRevisonsStable()
 		if err != nil {
```
Comment (xiang90) [inline tools/functional-tester/etcd-tester/checks.go]: if err != nil || !revStable {continue?}

File: tools/functional-tester/etcd-tester/checks.go [hunk 7]
```diff
@@ -39,39 +44,81 @@ func newHashChecker(hrg hashAndRevGetter) Checker { return &hashChecker{hrg} }
...
-			plog.Printf("#%d failed to get current revisions (%v)", i, err)
 			continue
 		}
-		if _, ok = getSameValue(revs); ok {
-			break
+		if revsStable {
```
Comment (xiang90) [inline tools/functional-tester/etcd-tester/checks.go]: if we change line 51, we do not need this checking here.

File: tools/functional-tester/etcd-tester/checks.go [hunk 8]
```diff
@@ -39,39 +44,81 @@ func newHashChecker(hrg hashAndRevGetter) Checker { return &hashChecker{hrg} }
...
+			plog.Printf("revisions are not stable: revisions [revisions: %v]", revs)
+			return false, nil
+		}
+		// sleep for N seconds. after that, check to make sure that revisions don't change
+		if i == 0 {
+			time.Sleep(sleepForStabliztionInSecond)
+		}
 	}
-	plog.Printf("all members are consistent with current revisions [revisions: %v]", revs)
+	plog.Printf("revisions are stable: revisions [revisions: %v]", revs)
```
Comment (xiang90) [inline tools/functional-tester/etcd-tester/checks.go]: dont we need to compare the 1st rev and the 2nd rev to ensure it is stable?

File: tools/functional-tester/etcd-tester/checks.go (hunk 5 above)
Comment (fanminshi) [inline tools/functional-tester/etcd-tester/checks.go]: sure

**Comment by fanminshi** [comment]
All fixed. PTAL @xiang90

**Comment by xiang90** [comment]
@fanminshi LGTM if this works well in your local testing env.

**Comment by fanminshi** [comment]
@xiang90 so far it works well in my testing env. We should deploy it to 5 nodes tester to see if it catches any failures.

**Pull request merged**